### ISSUE TITLE
mock(Class<T> classToMock, String name)  name was ignored

### ISSUE THREAD

**Issue by realvalkyrie** [mockito/mockito#1252]
i have run the test named "InjectionOfInlinedMockDeclarationIssueTest"
and try to modfied the name of the mock  failed with testing
, by it always passed test

`
@RunWith(MockitoJUnitRunner.class)
public class InjectionOfInlinedMockDeclarationIssueTest {

    @InjectMocks private Receiver receiver;
    @InjectMocks private Receiver spiedReceiver;

    private Antenna oldAntenna         = mock(Antenna.class);
    private Antenna satelliteAntenna = mock(Antenna.class);
    
   //opt1
    private Antenna antenna               = mock(Antenna.class,"dvbtAntenna");
   //opt2
    private Antenna antenna               = mock(Antenna.class,"testing");
   //opt3
    private Antenna antenna               = mock(Antenna.class);

    @Test
    public void named_mocks_should_be_resolved_with_their_name() throws Exception {
        //should not same
        assertSame(antenna    , receiver.dvbtAntenna);
        assertNotSame(antenna , oldAntenna);

        assertNotSame(receiver         , spiedReceiver);
        assertNotSame(oldAntenna       , satelliteAntenna);
        assertNotSame(oldAntenna       , antenna);
        assertNotSame(satelliteAntenna , antenna);
    }

    // note that static class is not private !!
    static class Receiver {
        Antenna oldAntenna;
        Antenna satelliteAntenna;
        Antenna dvbtAntenna;
    }
    private static class Antenna { }
}
`

leave the  opt1  or opt2 or opt3 alone ,  the  test was always passed .

plz fix this bug

**Comment by realvalkyrie** [comment]
opt1 will passed test 
opt2 and opt3 should failed with test

**Comment by TimvdLippe** [comment]
Yes this looks like a bug. Feel free to submit a PR, as the core team is busy with other tasks at the moment, we will likely not get to this issue soon.

**Comment by SeriyBg** [comment]
Hello, I would like to work on this issue, if it's ok :)

**Comment by SeriyBg** [comment]
Looks like the same issue is reproducible with the `@Mock` annotation.
Here is the test for it:
```java
    @InjectMocks
    private WithTwoFieldsOfSameType underTest;

    @Mock(name = "firstField")
    private Bean beanA;
    @Mock(name = "notExistingBean")
    private Bean beanB;
    .....

    @Test
    public void should_not_inject_mock_with_different_names() {
        assertNotSame(underTest.secondField, beanB);
    }

    private static class WithTwoFieldsOfSameType {
        private Bean firstField;
        private Bean secondField;
    }
    private interface Bean {}
```

**Comment by TimvdLippe** [comment]
@SeriyBg Sure go ahead :smile:

**Comment by SeriyBg** [comment]
@TimvdLippe I have done some research on this issue, so I want to confirm that the current behavior is wrong.
So it looks like on the injection of `firstField` is ambiguous by type so it considers the mock name. But when injecting `secondField ` the only available option is `beanB` as `beanA` was already injected by name. Is this behavior wrong or am I missing something?

**Comment by TimvdLippe** [comment]
@SeriyBg that sounds oddly familiar to #215. This bug should have been fixed already? 😮

**Comment by SeriyBg** [comment]
@TimvdLippe not exactly. The #215 describes the issue when you have multiple fields but only one mock. Here the behavior looks to work correctly. So the `beanA` is injected to the `firstField ` by name, and the `beanB` is injected to the `secondField` by type ignoring the name. This matches the documentation that says:

> mocks will first be resolved by type (if a single type match injection will happen regardless of the name)

Isn't it exactly what we have here?

**Comment by TimvdLippe** [comment]
Hm yes you are correct. Our injection mechanism is rather rudimentary and confusion like this issue makes us believe it is not something we should do.

In the end, it is probably working as intended, although unfortunate. I am hesitant to add even more logic to this component, as it is already very complicated. Modifying might mean we regress in other cases, meaning we are backwards incompatible.

**Comment by SeriyBg** [comment]
@TimvdLippe so you suggest not to change the current logic for with this issue, right?
(Sigh) I was hoping to get the first code contribution to mockito :pensive:

Thanks for your help here!

**Comment by TimvdLippe** [comment]
@SeriyBg I think this is working as intended, but maybe one of the other @mockito/core can chime in as well.

If you would like to help us out, we are eagerly looking for contributions! Check out this list of issues to find an easy one to start (or participate in the discussion): https://github.com/mockito/mockito/issues?q=is%3Aissue+is%3Aopen+label%3A%22please+contribute%22

**Comment by raphw** [comment]
Sorry this took so long. As a matter of fact, I would suggest that Mockito should throw an exception if mocks of the same type are in conflict to one another. This is always better then undefined behavior.

We could also add a boolean property to `@InjectMocks` that allows to fall back to the old behavior but in this case, I would actually suggest to just throw the exception as this behavior might just break by an implementation detail in the future anyways.
